File: sdk/typescript/src/credentials.ts
/**
 * Attaching an API key to every call.
 *
 * The key travels in call metadata rather than in a request body: a request
 * carrying its own answer to "who am I" has not been authenticated, it has been
 * asked politely (DESIGN.md §23).
 *
 * Metadata per call rather than composed channel credentials, because grpc-js
 * refuses to compose call credentials onto an insecure channel — it will not
 * help you send a secret in the clear. That refusal is right, and it is also why
 * this route exists: local development is plaintext, and the same code has to
 * work there and against TLS.
 */

import * as grpc from "@grpc/grpc-js";

/**
 * The key that was passed, or the one in the environment.
 *
 * Reading `PIPEMESH_API_KEY` means a key never has to be written into code that
 * gets committed.
 */
export function resolveKey(apiKey?: string): string | undefined {
  const fromEnvironment = process.env.PIPEMESH_API_KEY?.trim();
  return apiKey || fromEnvironment || undefined;
}

/**
 * The metadata every call carries — empty when there is no key.
 *
 * Warns when a key would travel over a plaintext connection. Not refused:
 * plaintext is right on a laptop, and refusing would make every development
 * setup stand up TLS first. But a key sent unencrypted to a real deployment is a
 * leaked key, and silence is how that happens.
 */
export function authMetadata(
  apiKey: string | undefined,
  secure: boolean,
): grpc.Metadata {
  const metadata = new grpc.Metadata();
  if (!apiKey) return metadata;

  if (!secure) {
    console.warn(
      "Sending an API key over a plaintext connection. That is fine locally and a leak " +
        "anywhere else; use a secure channel against a real deployment.",
    );
  }

  metadata.set("authorization", `Bearer ${apiKey}`);
  return metadata;
}

File: sdk/python/pipemesh/credentials.py
"""Attaching an API key to every call.

The key travels in call metadata, not in a request body: a request that carries
its own answer to "who am I" has not been authenticated, it has been asked
politely (DESIGN.md §23).
"""

from __future__ import annotations

import os
import warnings
from typing import Optional

import grpc

ENVIRONMENT_VARIABLE = "PIPEMESH_API_KEY"


def resolve(api_key: Optional[str]) -> Optional[str]:
    """The key that was passed, or the one in the environment."""
    if api_key:
        return api_key
    from_environment = os.environ.get(ENVIRONMENT_VARIABLE, "").strip()
    return from_environment or None


def channel_for(target: str, api_key: Optional[str]) -> grpc.Channel:
    """A plaintext channel, warning when a key would travel over it in the clear.

    Not refused: plaintext is right on a laptop, and refusing would make every
    development setup stand up TLS first. But a key sent unencrypted to a real
    deployment is a leaked key, and silence is how that happens.
    """
    if api_key:
        warnings.warn(
            "Sending an API key over a plaintext connection. That is fine locally and a "
            "leak anywhere else; use a secure channel against a real deployment.",
            stacklevel=3,
        )
    return grpc.insecure_channel(target)


def interceptors(api_key: Optional[str]):
    """Adds `authorization: Bearer <key>` to every call on the channel."""
    if not api_key:
        return ()
    header = ("authorization", f"Bearer {api_key}")
    return (
        _Unary(header),
        _UnaryStream(header),
        _StreamStream(header),
    )


class _Attaches:
    """Puts the header on a call, leaving whatever else was there alone."""

    def __init__(self, header) -> None:
        self._header = header

    def _with_header(self, client_call_details):
        metadata = list(client_call_details.metadata or [])
        metadata.append(self._header)
        return client_call_details._replace(metadata=metadata)


class _Unary(_Attaches, grpc.UnaryUnaryClientInterceptor):

    def intercept_unary_unary(self, continuation, client_call_details, request):
        return continuation(self._with_header(client_call_details), request)


class _UnaryStream(_Attaches, grpc.UnaryStreamClientInterceptor):

    def intercept_unary_stream(self, continuation, client_call_details, request):
        return continuation(self._with_header(client_call_details), request)


class _StreamStream(_Attaches, grpc.StreamStreamClientInterceptor):

    def intercept_stream_stream(self, continuation, client_call_details, request_iterator):
        return continuation(self._with_header(client_call_details), request_iterator)
